src/passes/LogFunction.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Instruments the Wasm binary with code to log execution at each function
+// entry with an index. This list of functions will be used when we split the
+// wasm later.
+//
+// The logging is performed by calling an ffi with an id for each
+// call site. You need to provide the import on the JS side.
+//
+
+#include "asm_v_wasm.h"
+#include "asmjs/shared-constants.h"
+#include "shared-constants.h"
+#include <pass.h>
+#include <wasm-builder.h>
+#include <wasm.h>
+
+namespace wasm {
+
+Name FNLOGGER("log_function");
+
+struct LogFunction : public WalkerPass<PostWalker<LogFunction>> {
+  void visitFunction(Function* curr) {
+    if (curr->imported()) {
```

Comment:
Since this doesn't really need to walk the full bodies of each function, you could make this a regular `Pass` and use `ModuleUtils::iterDefinedFunctions` (from module-utils.h) inside its `run` method rather than using `visitFunction`.

Reply:
> 
> 
> Since this doesn't really need to walk the full bodies of each function, you could make this a regular `Pass` and use `ModuleUtils::iterDefinedFunctions` (from module-utils.h) inside its `run` method rather than using `visitFunction`.

We will be instrumenting at the block level moving forward so is `ModuleUtils::iterDefinedFunctions` sufficient moving forward?
Also, we are using `visitModule` to register the JS import for now. We can rework this when you are done with handcrafting the Wasm function to create the profile data?